Title: <center>Jeffrey W. Grover</center>
Date: 2019-06-19
save_as: pages/cv.html

<center>Tucson, AZ</center>
<center>**E-mail:** [jeffrey.w.grover@gmail.com](mailto:jeffrey.w.grover@gmail.com)</center>

#### **Education**
**2020 – Ph.D. – Molecular and Cellular Biology**<br>
**The University of Arizona**<br>
Advisor: Dr. Rebecca Mosher<br>
Dissertation: Investigating the Role of RNA-directed DNA Methylation in Plant
Reproduction and Development of Computational Resources Enbaling Epigenomic
Research

**2015 – Professional Science Masters (PSM) – Applied Biosciences**<br>
**Specialization: Molecular and Cellular Biology**<br>
**The University of Arizona**<br>
Advisor: Dr. Rebecca Mosher<br>
Thesis: Investigation of Nuclear-Cytoplasmic Shuttling of AGO4 and its Effect on
RNA-directed DNA Methylation in *Arabidopsis thaliana*

**2011 – B.S. – Biochemistry and Molecular Biology**<br>
**Additional Major: Chemistry**<br>
**Michigan State University**

#### **Professional/Research Experience**
**2020-Present - Genomics Scientist, Seven Bridges**<br>
**Boston, MA**
**Public Programs Division**

- Facilitated multi-omics data analysis from multi-site sequencing pilot experiments
- Coordinated availability of sequencing data and metadata on the [Cancer Genoimcs Cloud (CGC)](http://www.cancergenomicscloud.org/)
- Developed and validated bioinformatics workflows on the CGC using Common Workflow Language (CWL)
- Enabled discoverability of sequencing data and metadata with R Shiny-bashed dashboards
- Advised on data visualization techniques for multi-omics data

**2014-2020 – Graduate Assistant, The University of Arizona**<br>
**Tucson, AZ**
**Lab: Dr. Rebecca Mosher - Molecular and Cellular Biology & The School of Plant Sciences**

- Integrated small RNA seq, whole-genome bisulfite sequencing, and RNAseq analyses.
- Developed reproducible analysis pipelines with workflow managers and containers.
- Created visualizations from genomics data for publication and presentation.
- Administered a Linux-based bioinformatics server for multiple users.
- Developed web-based resources for the analysis of high throughput sequencing data.
- Constructed sequencing libraries for Illumina platform.
- Designed and performed methylation-sensitive quantitative PCR assays.
- Assessed protein localization using confocal microscopy and western blot.

**2011-2013 – Research Associate, MPI Research Inc.**<br>
**Mattawan, MI**
**Cellular, Molecular, and Biochemical Sciences Department**

- Performed qPCR-based biodistribution assays and expression profiling by qPCR arrays.
- Performed high throughput DNA/RNA isolation with laboratory automation.
- Conducted assays in compliance with the FDA regulations (GLPs).
- Authored GLP-compliant experimental protocols and reviewed data for GLP compliance.

**2007-2011 – Undergraduate Research Assistant, Michigan State University**<br>
**East Lansing, MI**
**Lab: Dr. Robert Last - Department of Biochemistry and Molecular Biology**

- Assayed Arabidopsis thaliana mutants for lipid content by gas chromatography.
- Mapped lipid phenotypic traits to candidate genomic loci using microarrays.
- Contributed data to the Chloroplast Phenomics Database (plastid.msu.edu).

#### **Technical Skills**
**Computational**

- Python programming – Workflow management and data processing
- R programming – Visualization, statistical analysis, and data processing
- BASH scripting – Workflow management and data processing
- Sequencing analysis – mRNAseq, small RNAseq, and bisulfite sequencing
- Linux system administration – Software management and troubleshooting user issues
- High performance computing – Deploying workflows
- Version control – git and GitHub

**Molecular Biology**

- Quantitative PCR – Data analysis and experimental design
- Sequencing library preparation and QC – Illumina platform
- Protein analysis – SDS-PAGE and western blotting

#### **Mentorship**
**2019-2020 – Mosher Lab, University of Arizona**<br>
Mentored an undergraduate student working on a project with next-generation
sequencing data analysis and reproducible analysis pipelines.

**2017-2018 – Mosher Lab, University of Arizona**<br>
Mentored a high school student who completed a senior project on plant
epigenetics.

**2014-2015 – Mosher Lab, University of Arizona**<br>
Mentored a high school student who completed a research project in epigenetics
and won multiple awards at regional and national science fairs.

#### **Teaching Experience**
**2019 – Teaching Assistant, University of Arizona**<br>
MCB301: Molecular Basis of Life (3rd Year: Biochemistry and Molecular Biology)

**2017 – Teaching Assistant, University of Arizona**<br>
MCB304: Molecular Genetics (3rd/4th Year: Genetics)

#### **Broader Impacts**

**Tool Development**

**[WGBS Snakemake Pipeleine](https://github.com/groverj3/wgbs_snakemake)**<br>
An automated whole-genome bisulfite sequencing pipeline. Available on
[Github](https://github.com/groverj3/wgbs_snakemake) and as a [Docker
container](https://hub.docker.com/r/groverj3/mosher_lab_wgbs). Compatible with
Singularity for deployment on HPC systems.

**[sRNAseq Snakemake Pipeleine](https://github.com/boseHere/sRNA_snakemake_workflow)**<br>
An automated small-RNA sequencing pipeline. Available on
[Github](https://github.com/boseHere/sRNA_snakemake_workflow) and as a [Docker
container](https://hub.docker.com/r/bose1/mosher_lab_srna). Compatible with
Singularity for deployment on HPC systems.

- Supervised its creation by undergraduate student, Maya Bose.
- The pipeline is based on the methods from [Grover et al. 2018](https://doi.org/10.1111/tpj.13910)

**[LoadExp+](https://genomevolution.org/wiki/index.php/LoadExp%2B)**<br>
Web-based analysis pipelines next-generation sequencing experiments
[CoGe](https://genomevolution.org). Integrated with the
[CyVerse](www.cyverse.org) datastore.

**Publicly Available Data**

Small RNA Sequencing from *B. rapa* ovules and leaves - R-o-18, *nrpd1*, *nrpe1*,
and *rdr2* mutants<br> 
NCBI SRA Accession: [SRP114437](https://www.ncbi.nlm.nih.gov/sra/SRP114437)

Small RNA Sequencing from *B. rapa* R-o-18xR500 reciprocal cross seeds<br>
NCBI SRA Accession: [SRP114469](https://www.ncbi.nlm.nih.gov/sra/SRP114469)

RNA Sequencing from *B. rapa* R-o-18 and *nrpd1* mutant ovules and seeds<br>
NCBI SRA Accession: [SRP132223](https://www.ncbi.nlm.nih.gov/sra/SRP132223)

*B. rapa* Seed Epigenome Profiling<br>
- Tissues: Whole seeds, dissected seed tissues, ovules, and leaves<br>
- Genotypes: RdDM mutants, wild-types of two *B. rapa* varieties, and reciprocal crosses<br>
- Data Types: Small RNAseq, whole-genome bisulfite sequencing, and mRNA-seq<br>
- NCBI SRA Accession: [SRP229027](https://www.ncbi.nlm.nih.gov/sra/SRP229027)

#### **Conference Presentations**

**Oral Presentations**

**2018 – The University of Arizona Joint Biology Retreat**<br>
Dissection of the Spatiotemporal Dynamics of RNA-directed DNA Methylation During
Seed Development

**2017 – Plant and Animal Genome Conference, San Diego, CA**<br>
User-friendly Whole Genome DNA Methylation Analysis with FlowGe

**Poster Presentations**

**2019 – Plant and Animal Genome Conference, San Diego, CA**<br>
Spatiotemporal Dynamics of Small RNA Accumulation During Seed Development in
*Brassica rapa*

**2017 – The University of Arizona Joint Biology Retreat**<br>
PolIV-dependent siRNAs from Maternal Somatic Tissue Are Required for Seed
Development in *Brassica rapa*

**2016 – Arizona RNA Symposium**<br>
AGO4 Nuclear Import and Export Impact RNA-directed DNA Methylation

**2016 – The University of Arizona Joint Biology Retreat**<br>
AGO4 Nuclear Import and Export Impact RNA-directed DNA Methylation

**2015 – The University of Arizona School of Plant Sciences Departmental Retreat**<br>
Investigation of Nuclear-Cytoplasmic Shuttling of AGO4 and its Effect on 
RNA-directed DNA Methylation in *Arabidopsis thaliana*

#### **Publications**
**Grover JW**, Burgess D, Kendall T, Baten A, Pokhrel S, King GJ, Meyers BC, Freeling M,
Mosher RA. Abundant expression of maternal siRNAs is a conserved feature of seed
development. bioRxiv. 2019. [https://doi.org/10.1101/866806](https://doi.org/10.1101/866806)<br>
- Preprint available before peer review and final publication


**Grover JW\***, Kendall T\*, Baten A, Burgess D, Freeling M, King GJ, and Mosher RA. *Maternal
components of RNA‐directed DNA methylation are required for seed development in Brassica
rapa.* The Plant Journal. 2018. [https://doi.org/10.1111/tpj.13910](https://doi.org/10.1111/tpj.13910)

- \* These authors contributed equally to this work.<br>
- Chosen as the cover article by The Plant Journal.<br>
- [Covered by a research highlight in the Plant Journal.](https://doi.org/10.1111/tpj.13935)<br>
- [Mentioned in Plantae’s “What We’re Reading” blog.](https://plantae.org/what-were-reading-april-20th/)

**Grover JW**, Bomhoff M, Davey S, Gregory BD, Mosher RA, Lyons E. *CoGe LoadExp+:
A web‐based suite that integrates next‐generation sequencing data analysis workflows and
visualization*. Plant Direct. 2017. [https://doi.org/10.1002/pld3.8](https://doi.org/10.1002/pld3.8)

#### **References**
Available upon request
