Title: <center>Jeffrey W. Grover</center>
Date: 2024-04-20
save_as: pages/cv.html

<center>Boston, MA</center>
<center>**E-mail:** [jeffrey.w.grover@protonmail.com](mailto:jeffrey.w.grover@protonmail.com)</center>

## **Education**
**2020 – Ph.D. – Molecular and Cellular Biology**<br>
**The University of Arizona**<br>
Advisor: Dr. Rebecca Mosher<br>
Dissertation: Investigating the Role of RNA-directed DNA Methylation in Plant
Reproduction and Development of Computational Resources Enbaling Epigenomic
Research

**2015 – Professional Science Masters (PSM) – Applied Biosciences**<br>
**Specialization: Molecular and Cellular Biology**<br>
**The University of Arizona**<br>
Advisor: Dr. Rebecca Mosher<br>
Thesis: Investigation of Nuclear-Cytoplasmic Shuttling of AGO4 and its Effect on
RNA-directed DNA Methylation in *Arabidopsis thaliana*

**2011 – B.S. – Biochemistry and Molecular Biology**<br>
**Additional Major: Chemistry**<br>
**Michigan State University**

## **Professional/Research Experience**
### **Entrada Therapeutics - Boston, MA**
**2024-Present – Senior Scientist I - NGS & Bioinformatics**<br>
**Quantitative Biology Division**

- Acted as subject matter expert across therapeutic programs to assess feasibility and actionability of NGS, bioinformatics, and data science project proposals.
- Led the transcriptomic characterization of the toxicity profile of a novel class of protein-conjugated oligonucleotide therapeutics providing valuable data to candidate selection.
- Standardized Entrada’s bulk and single-cell RNAseq workflows allowing valid comparisons across studies and the collection of a large library of NGS data.
- Implemented methods for pseudobulk differential expression in Entrada’s single-cell RNAseq analysis.
- Designed in-vitro and in-vivo experiments to assess organ-specific toxicity of drug candidates by RNAseq and single-cell RNAseq.
- Developed a strategic plan for computational infrastructure and data storage in coordination with IT and Quality Assurance.

**2022-2024 – Scientist II - NGS & Bioinformatics**<br>
**Discovery Biology and Quantitative Biology Divisions**

- Entrada's first bioinformatics and NGS hire, tasked with building-out bioinformatics and NGS capabilities from scratch.
- Collaborated across company divisions to leverage bioinformatics, data science, and NGS on a variety of therapeutic programs and for discovery efforts.
- Coordinated with IT to build and manage hybrid cloud bioinformatics computing and data storage infrastructure.
- Co-led NGS laboratory and instrumentation set-up.
- Managed the company’s Illumina sequencing instruments and NGS data/metadata.
- Collaboratively designed and executed RNAseq-based studies for off-target effect of a clinical candidate drug on tight timelines that contributed data to an IND filing.
- Analyzed NGS data for differential gene expression, alternative splicing, gene editing, single-cell RNAseq, and more.
- Developed automated bioinformatics workflows with NextFlow.
- Instituted best practices for bioinformatics workflow development and deployment.
- Authored SOPs to ensure reproducibility of computational analyses and data integrity.
- Supervised and trained junior staff in bioinformatics, data science, and NGS lab work.
- Presented data to internal stakeholders and across differing levels of technical expertise.

### **Seven Bridges Genomics - Boston, MA**
**2020-2022 - Genomics Scientist**<br>
**Public Programs/Seven Bridges Labs Division**

- Built reproducible bioinformatics analysis workflows with cloud computing, Docker containers, the Seven Bridges Platform, and the Common Workflow Language (CWL).
- Analyzed patient-derived xenograft genomics data (transcriptomics and whole exome) for the NCI’s PDXNet Consortium (https://www.pdxnetwork.org/).
- Managed access to PDXNet data through the Seven Bridges/NCI Cancer Genomics Cloud (https://www.cancergenomicscloud.org/)
- Contributed to development of the PDXNet Portal (https://portal.pdxnetwork.org/) using R Shiny.
- Piloted internal data science and genomics analysis initiatives.
- Consulted on genomics and epigenomics analysis initiatives.
- Contributed to Seven Bridges Platform development, proposing new features, bug fixes, and strategies to capture new bioinformatics use cases and users.
- Participated in outreach to users assisting solving technical and bioinformatics problems
- Hosted webinars on bioinformatics topics and Seven Bridges platform/tooling features

### **The University of Arizona - Tucson, AZ**
**2014-2020 – Graduate Student**<br>
**Lab: Dr. Rebecca Mosher - Molecular and Cellular Biology & The School of Plant Sciences**

- Integrated small RNA seq, whole-genome bisulfite sequencing, and RNAseq analyses.
- Developed reproducible analysis pipelines with workflow managers and containers.
- Created visualizations from genomics data for publication and presentation.
- Administered a Linux-based bioinformatics server for multiple users.
- Developed web-based resources for the analysis of high throughput sequencing data.
- Constructed sequencing libraries for Illumina platform.
- Designed and performed methylation-sensitive quantitative PCR assays.
- Assessed protein localization using confocal microscopy and western blot.

### **MPI Research, Inc. - Mattawan, MI**
**2011-2013 – Research Associate**<br>
**Cellular, Molecular, and Biochemical Sciences Department**

- Performed qPCR-based biodistribution assays and expression profiling by qPCR arrays.
- Performed high throughput DNA/RNA isolation with laboratory automation.
- Conducted assays in compliance with the FDA regulations (GLPs).
- Authored GLP-compliant experimental protocols and reviewed data for GLP compliance.

### **Michigan State University - East Lansing, MI**
**2007-2011 – Undergraduate Research Assistant**<br>
**Lab: Dr. Robert Last - Department of Biochemistry and Molecular Biology**

- Assayed Arabidopsis thaliana mutants for lipid content by gas chromatography.
- Mapped lipid phenotypic traits to candidate genomic loci using microarrays.
- Contributed data to the Chloroplast Phenomics Database (plastid.msu.edu).

## **Technical Skills**
### **Computational**

- Genomics Workflows – Snakemake, Nextflow, Common Workflow Language
- Python programming – Workflow management and data processing
- R programming – Visualization, statistical analysis, and data processing
- Data Science – Clustering, dimensionality reduction, large dataset manipulation, SQL, application of standard machine learning algorithms
- BASH scripting – Workflow management and data processing
- Computational notebooks – Jupyter, R Markdown
- NGS experimental design – RNAseq, targeted RNAseq, single-cell/nucleus RNAseq, small RNAseq, whole-exome sequencing, and bisulfite sequencing
- Bioinformatics analysis – Differential gene expression, variant calling, DNA methylation, scRNAseq cell typing, scRNAseq pseudobulk differential expression
- Linux system administration – Software and operating system management
- High performance computing – Job submission (SLURM), software deployment with containers (Docker, Podman, Singularity, Apptainer, et al.)
- Version control – git and GitHub/GitLab for genomics workflow development and computational notebook management
- Cloud computing - Running HPC workloads and managing storage on public clouds (AWS, GCP, Azure, etc.)

### **Molecular Biology**

- Sequencing library preparation and QC – Illumina platform
- Protein analysis – SDS-PAGE and western blotting
- Nucleic acid techniques – RNA/DNA isolation, gel electrophoresis, PCR, RT-PCR, qPCR

## **Mentorship**
**2022-2024 – Entrada Therapeutics**
- Trained and mentored a research associate on bioinformatics and NGS methods
resulting in increases in the Quantitative Biology department’s bioinformatics analysis

**2019-2020 – Mosher Lab, University of Arizona**<br>
Mentored an undergraduate student working on a project with next-generation
sequencing data analysis and reproducible analysis pipelines.

**2017-2018 – Mosher Lab, University of Arizona**<br>
Mentored a high school student who completed a senior project on plant
epigenetics.

**2014-2015 – Mosher Lab, University of Arizona**<br>
Mentored a high school student who completed a research project in epigenetics
and won multiple awards at regional and national science fairs.

## **Teaching Experience**
**2019 – Teaching Assistant, University of Arizona**<br>
MCB301: Molecular Basis of Life (3rd Year: Biochemistry and Molecular Biology)

**2017 – Teaching Assistant, University of Arizona**<br>
MCB304: Molecular Genetics (3rd/4th Year: Genetics)

## **Broader Impacts**
### **Tool Development**

**Epigenomics on the Cancer Genomics Cloud (CGC)**<br>
Oversaw the addition of [ChIPseq](https://cgc.sbgenomics.com/public/apps#admin/sbg-public-data/encode-chip-seq-pipeline-2/)
and [ATACseq](https://cgc.sbgenomics.com/public/apps#admin/sbg-public-data/encode-atac-seq-pipeline-1-9-2/)
workflows to the CGC in collaboration with Seven Bridges Bioinformatics team in Europe.

- Provided technical and scientific consulting regarding their implementation.
- Publicized these workflows in a talk at the [CGC Summer Symposium 2021](https://www.cancergenomicscloud.org/events/2021/8/18/cgc-summer-symposium-2021).

**[PDXNet Public Data Portal](https://portal.pdxnetwork.org/)**<br>
Guided development of the PDXNet Public Data Portal (https://portal.pdxnetwork.org/).
The PDXNet Portal is an R Shiny-based web application where analyses performed
on data generated by the PDXNet initiative can be visualized and the consortium
data is made available to the public.

**[WGBS Snakemake Pipeleine](https://github.com/groverj3/wgbs_snakemake)**<br>
An automated whole-genome bisulfite sequencing pipeline. Available on
[Github](https://github.com/groverj3/wgbs_snakemake) and as a [Docker
container](https://hub.docker.com/r/groverj3/mosher_lab_wgbs). Compatible with
Singularity for deployment on HPC systems.

**[sRNAseq Snakemake Pipeleine](https://github.com/boseHere/sRNA_snakemake_workflow)**<br>
An automated small-RNA sequencing pipeline. Available on
[Github](https://github.com/boseHere/sRNA_snakemake_workflow) and as a [Docker
container](https://hub.docker.com/r/bose1/mosher_lab_srna). Compatible with
Singularity for deployment on HPC systems.

- Supervised its creation by undergraduate student, Maya Bose.
- The pipeline is based on the methods from [Grover et al. 2018](https://doi.org/10.1111/tpj.13910)

**[LoadExp+](https://genomevolution.org/wiki/index.php/LoadExp%2B)**<br>
Web-based analysis pipelines next-generation sequencing experiments
[CoGe](https://genomevolution.org). Integrated with the
[CyVerse](www.cyverse.org) datastore.

### **Publicly Available Data**

PDXNet RNAseq and Whole-exome Sequencing<br>
- Available through https://portal.pdxnetwork.org/<br>
- Processed over 80TB of raw sequencing data

Small RNA Sequencing from *B. rapa* ovules and leaves - R-o-18, *nrpd1*, *nrpe1*,
and *rdr2* mutants<br> 
NCBI SRA Accession: [SRP114437](https://www.ncbi.nlm.nih.gov/sra/SRP114437)

Small RNA Sequencing from *B. rapa* R-o-18xR500 reciprocal cross seeds<br>
NCBI SRA Accession: [SRP114469](https://www.ncbi.nlm.nih.gov/sra/SRP114469)

RNA Sequencing from *B. rapa* R-o-18 and *nrpd1* mutant ovules and seeds<br>
NCBI SRA Accession: [SRP132223](https://www.ncbi.nlm.nih.gov/sra/SRP132223)

*B. rapa* Seed Epigenome Profiling<br>
- Tissues: Whole seeds, dissected seed tissues, ovules, and leaves<br>
- Genotypes: RdDM mutants, wild-types of two *B. rapa* varieties, and reciprocal crosses<br>
- Data Types: Small RNAseq, whole-genome bisulfite sequencing, and mRNA-seq<br>
- NCBI SRA Accession: [SRP229027](https://www.ncbi.nlm.nih.gov/sra/SRP229027)

## **Conference Presentations**
### **Oral Presentations**

**2021 – NCI Containers and Workflows Interest Group Webinar Series**<br>
[Developing Scalable Bioinformatics Workflows on the Cancer Genomics Cloud](https://cbiit.webex.com/recordingservice/sites/cbiit/recording/1e77619d2621103aaed50050568c91b6/playback)

**2021 – The Cancer Genomics Cloud Summer Symposium**<br>
[Epigenomics On the Cancer Genomics Cloud: Enabling Scalable Analysis Using Cloud Computing](https://www.youtube.com/watch?v=qebAyqRAgrk)

- Attended by over 100 CGC users


**2018 – The University of Arizona Joint Biology Retreat**<br>
Dissection of the Spatiotemporal Dynamics of RNA-directed DNA Methylation During
Seed Development

**2017 – Plant and Animal Genome Conference, San Diego, CA**<br>
User-friendly Whole Genome DNA Methylation Analysis with FlowGe

### **Poster Presentations**

**2021 – Bioinformatics Open Source Conference, Online**<br>
Linking Standards through the BioCompute App to Support Workflow Interpretability and Execution

**2019 – Plant and Animal Genome Conference, San Diego, CA**<br>
Spatiotemporal Dynamics of Small RNA Accumulation During Seed Development in
*Brassica rapa*

**2017 – The University of Arizona Joint Biology Retreat**<br>
PolIV-dependent siRNAs from Maternal Somatic Tissue Are Required for Seed
Development in *Brassica rapa*

**2016 – Arizona RNA Symposium**<br>
AGO4 Nuclear Import and Export Impact RNA-directed DNA Methylation

**2016 – The University of Arizona Joint Biology Retreat**<br>
AGO4 Nuclear Import and Export Impact RNA-directed DNA Methylation

**2015 – The University of Arizona School of Plant Sciences Departmental Retreat**<br>
Investigation of Nuclear-Cytoplasmic Shuttling of AGO4 and its Effect on 
RNA-directed DNA Methylation in *Arabidopsis thaliana*

## **Publications**
Burgess D, Chow HT, **Grover JW**, Freeling M, Mosher RA. Ovule siRNAs methylate protein-
coding genes in trans. The Plant Cell. 2022;34(10):3647-3664.
[https://doi.org/10.1093/plcell/koac197](https://doi.org/10.1093/plcell/koac197)

Koc S, Lloyd MW, **Grover JW**, Xiao N, Seepo S, Subramanian SL, Ray M, Frech C,
DiGiovanna J, Webster P, Neuhauser S, Srivastava A, Woo XY, Sanderson BJ,
White B, Lott P, Dobrolecki LE, Dowst H, PDXNet Consortium, Evrard YA,
Wallace TA, Moscow JA, Doroshow JH, Mitsiades N, Kaochar S, Pan CX, Chen MS,
Carvajal-Carmona L, Welm AL, Welm BE, Lewis MT, Govindan R, Ding L, Li S,
Herlyn M, Davies MA, Roth J, Meric-Bernstam F, Robinson PN, Bult CJ,
Davis-Dusenbery B, Dean II DA, Chuang JH. *PDXNet portal: patient-derived
Xenograft model, data, workflow and tool discovery*. NAR Cancer. 2022.
[https://doi.org/10.1093/narcan/zcac014](https://doi.org/10.1093/narcan/zcac014)

Becktel DA, Zbesko JC, B. Frye JB, Chung AG, Hayes M, Calderon K, **Grover JW**,
Li A, Garcia FG, Tavera-Garcia MA, Schnellmann RG, Wu HJ, Nguyen TV, Doyle KP.
*Repeated administration of 2-hydroxypropyl-β-cyclodextrin (HPβCD) attenuates the
chronic inflammatory response to experimental stroke. Journal of Neuroscience*.
2021. [https://doi.org/10.1523/JNEUROSCI.0933-21.2021](https://doi.org/10.1523/JNEUROSCI.0933-21.2021)

Chakraborty T, Kendall T, **Grover JW**, Mosher RA. *Embryo CHH hypermethylation
is mediated by RdDM and is autonomously directed in Brassica rapa*. Genome
Biology. 2021. [https://doi.org/10.1186/s13059-021-02358-3](https://doi.org/10.1186/s13059-021-02358-3)

**Grover JW**, Burgess D, Kendall T, Baten A, Pokhrel S, King GJ, Meyers BC, Freeling M,
Mosher RA. *Abundant expression of maternal siRNAs is a conserved feature of seed
development*. PNAS. 2020. [https://doi.org/10.1073/pnas.2001332117](https://doi.org/10.1073/pnas.2001332117)

**Grover JW\***, Kendall T\*, Baten A, Burgess D, Freeling M, King GJ, and Mosher RA. *Maternal
components of RNA‐directed DNA methylation are required for seed development in Brassica
rapa.* The Plant Journal. 2018. [https://doi.org/10.1111/tpj.13910](https://doi.org/10.1111/tpj.13910)

- \* These authors contributed equally to this work.<br>
- Chosen as the cover article by The Plant Journal.<br>
- [Covered by a research highlight in the Plant Journal.](https://doi.org/10.1111/tpj.13935)<br>
- [Mentioned in Plantae’s “What We’re Reading” blog.](https://plantae.org/what-were-reading-april-20th/)

**Grover JW**, Bomhoff M, Davey S, Gregory BD, Mosher RA, Lyons E. *CoGe LoadExp+:
A web‐based suite that integrates next‐generation sequencing data analysis workflows and
visualization*. Plant Direct. 2017. [https://doi.org/10.1002/pld3.8](https://doi.org/10.1002/pld3.8)

#### **References**
Available upon request
