
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="HandheldFriendly" content="True" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta name="robots" content="" />

  <link href="https://fonts.googleapis.com/css?family=Source+Code+Pro|Source+Sans+Pro:300,400,400i,700" rel="stylesheet">

    <link rel="stylesheet" type="text/css" href="https://groverj3.github.io/theme/stylesheet/style.min.css">

  <link rel="stylesheet" type="text/css" href="https://groverj3.github.io/theme/pygments/paraiso-dark.min.css">
  <link rel="stylesheet" type="text/css" href="https://groverj3.github.io/theme/font-awesome/css/fontawesome.css">
  <link rel="stylesheet" type="text/css" href="https://groverj3.github.io/theme/font-awesome/css/brands.css">
  <link rel="stylesheet" type="text/css" href="https://groverj3.github.io/theme/font-awesome/css/solid.css">


    <link href="https://groverj3.github.io/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="Jeff Grover. Bioinformatician. Atom">





<meta name="author" content="Jeffrey Grover" />
<meta name="description" content="Commonly, in bioinformatics we&#39;re in the business of determining whether something, be it gene expression, or DNA methylation, or splicing, etc. is different between multiple conditions. Typically this would be done by comparing those data and using some kind of statistical test. However, with the continued advances in sequencing technologies …" />
<meta name="keywords" content="bioinformatics, data-visualization">

<meta property="og:site_name" content="Jeff Grover. Bioinformatician."/>
<meta property="og:title" content="Making Better Metaplots With ggplot, Part 1"/>
<meta property="og:description" content="Commonly, in bioinformatics we&#39;re in the business of determining whether something, be it gene expression, or DNA methylation, or splicing, etc. is different between multiple conditions. Typically this would be done by comparing those data and using some kind of statistical test. However, with the continued advances in sequencing technologies …"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:url" content="https://groverj3.github.io/articles/2019-06-27_making-better-metaplots-with-ggplot-part-1.html"/>
<meta property="og:type" content="article"/>
<meta property="article:published_time" content="2019-06-27 00:00:00-07:00"/>
<meta property="article:modified_time" content=""/>
<meta property="article:author" content="https://groverj3.github.io/author/jeffrey-grover.html">
<meta property="article:section" content="how-to"/>
<meta property="article:tag" content="bioinformatics"/>
<meta property="article:tag" content="data-visualization"/>
<meta property="og:image" content="/images/jeff.jpg">

  <title>Jeff Grover. Bioinformatician. &ndash; Making Better Metaplots With ggplot, Part 1</title>

</head>
<body>
  <aside>
    <div>
      <a href="https://groverj3.github.io">
        <img src="/images/jeff.jpg" alt="Jeffrey Grover" title="Jeffrey Grover">
      </a>
      <h1><a href="https://groverj3.github.io">Jeffrey Grover</a></h1>

<p>Bioinformatician<br>
Genomics Scientist @ <a href="https://www.sevenbridges.com/" target="_blank">Seven Bridges</a></p>
      <nav>
        <ul class="list">

          <li><a href="/pages/about.html">About</a></li>
          <li><a href="/blog.html">Blog</a></li>
          <li><a href="/pages/cv.html">CV</a></li>
          <li><a href="/pages/projects.html">Projects</a></li>
        </ul>
      </nav>

      <ul class="social">
          <li>
            <a  class="sc-envelope" href="mailto:jeffrey.w.grover@gmail.com" target="_blank">
            <i class="fas fa-envelope">
            </i>
          </a></li>
          <li>
            <a  class="sc-linkedin" href="https://www.linkedin.com/in/jeffreygrover/" target="_blank">
            <i class="fab fa-linkedin">
            </i>
          </a></li>
          <li>
            <a  class="sc-github" href="https://github.com/groverj3" target="_blank">
            <i class="fab fa-github">
            </i>
          </a></li>
      </ul>
    </div>


  </aside>
  <main>

    <nav>
      <a href="https://groverj3.github.io">    Home
</a>

      <a href="/pages/about.html">About</a>
      <a href="/blog.html">Blog</a>
      <a href="/categories.html">Categories</a>
      <a href="/tags.html">Tags</a>

      <a href="https://groverj3.github.io/feeds/all.atom.xml">    Atom
</a>

    </nav>

<article class="single">
  <header>
      
    <h1 id="making-better-metaplots-with-ggplot-part-1">Making Better Metaplots With ggplot, Part 1</h1>
    <p>
          Posted on Thu 27 June 2019 in <a href="https://groverj3.github.io/category/how-to.html">how-to</a>


    </p>
  </header>


  <div>
    <p>Commonly, in bioinformatics we're in the business of determining whether
something, be it gene expression, or DNA methylation, or splicing, etc. is
different between multiple conditions. Typically this would be done by comparing
those data and using some kind of statistical test. However, with the continued
advances in sequencing technologies generating greater read depth, and these
technologies becoming more available to researchers we can also look at
genome-scale data in other ways. Testing purely on count or score data does not
inform one of the positional information associated with that data.</p>
<p>To look at the positional context associated with genomics data we have several
options. One common way is a visualization that's often referred to as a
"metaplot" or "metagene plot." These plots are similar to the TSS or "peak" plots
commonly used to visualize chip-seq or similar data. In a metaplot the entire
length of a feature is scaled such that each feature now is composed of the same
number of "bins" of data. This allows one to visualize the data associated with
these features across their entire length. There are existing software packages
that can make these plots without too much trouble such as 
<a href="https://deeptools.readthedocs.io/en/stable/">Deeptools</a> or the
<a href="https://bioconductor.org/packages/release/bioc/html/genomation.html">Genomation</a>
R library. In particular, I find Deeptools to be a great software package, and it
makes some wonderful visualizations that would be a pain to make yourself.
Genomation requires one to be very familiar with R since it isn't a standalone
program. Deeptools is easier to use but its metaplots leave something to be
desired:</p>
<p><center>
<img src="/figures/2019-06-27_making-better-metaplots-with-ggplot-part-1/deeptools_example_meta.png", style="width:460px;height:275px;">
</center></p>
<p>I like the control I have over all plot elements and the professional look that
<a href="https://ggplot2.tidyverse.org/">ggplot</a> affords. I use it for most of my data
visualization needs. So, I figured, why not make something prettier with it?</p>
<h3>1. Install Required Packages</h3>
<p>This guide will use Deeptools, a Python package with a ton of functionality that
you can play around with later, and ggplot2 from the
<a href="https://www.tidyverse.org/">tidyverse</a>. The tidyverse is a collection of R
libraries designed by Hadley Wickham that make data science a snap. You can
install them as follows in a terminal:</p>
<div class="highlight"><pre><span></span><code>pip install --user deeptools
</code></pre></div>


<p>Launch the R interpreter by typing <code>R</code> and then:</p>
<div class="highlight"><pre><span></span><code><span class="nf">install.packages</span><span class="p">(</span><span class="s">&#39;tidyverse&#39;</span><span class="p">)</span>
</code></pre></div>


<p>I recommend installing them into a user-specific library by either the <code>--user</code>
flag for pip or setting up a .Renviron file with a path to a local library. You
can learn how to do that in my
<a href="/articles/2019-06-25/managing-software-on-a-multiuser-linux-system.html">previous post</a>.
You're also going to need <a href="https://samtools.github.io/">samtools</a>. Feel free to
use the package manager of your choice if conda is more your jam.</p>
<h3>2. Generate the Data Table With Deeptools</h3>
<p>Now that you've got the software installed you'll need to generate per-position
"score" information. If this is expression data or similar you can use Deeptools
again. But you should be able to use other inputs to the later steps as well. If
using expression data you can use your bam file you can use Deeptools'
bamCoverage tool. First, you need to index the alignment .bam file:</p>
<div class="highlight"><pre><span></span><code>samtools index <span class="si">${</span><span class="nv">input_bam</span><span class="si">}</span> <span class="si">${</span><span class="nv">input_bam</span><span class="p">%.bam</span><span class="si">}</span>
</code></pre></div>


<p><code>${}</code> is the syntax for using a previously declared variable in BASH and I'll use
that kind of representation throughout for places where values should be
specified.</p>
<p>Now that you have that out of the way. Your first step is to generate a coverage
file in bigWig format. This is a binary format but contains similar data to a
<a href="https://genome.ucsc.edu/goldenPath/help/bedgraph.html">bedGraph</a>. You can use
the bamCoverage tool:</p>
<div class="highlight"><pre><span></span><code>bamCoverage <span class="se">\</span>
    -p <span class="si">${</span><span class="nv">threads_to_use</span><span class="si">}</span> <span class="se">\</span>
    --binSize <span class="m">1</span> <span class="se">\</span>
    --normalizeUsing <span class="si">${</span><span class="nv">RPKM</span><span class="p">|CPM|BPM|RPGC</span><span class="si">}</span> <span class="se">\</span>
    --outFileFormat bigwig <span class="se">\</span>
    --bam <span class="si">${</span><span class="nv">input_bam</span><span class="si">}</span> <span class="se">\</span>
    --outFileName <span class="si">${</span><span class="nv">input_bam</span><span class="p">%.bam</span><span class="si">}</span>.bigWig
</code></pre></div>


<p>A <code>--binsize</code> of 1 will just generate per-base converage. This may be slow, and
you could increase the value if you wish. There are also other ways of generating
coverage/depth such as <a href="https://github.com/brentp/mosdepth">mosdepth</a> (a great
tool by Brent Pedersen). This comes with Deeptools though, and is easy to get
running. The <code>--normalizeUsing</code> option will let you normalize the coverage by
several methods, which is particularly useful for plotting multiple experiments
together at the end.</p>
<p>Next, you'll need to generate a score matrix. In other words, a matrix of
coverages or other values of interest. This step can be done on any score data in
a bedGraph/bigWig file, even if you did not generate it with Deeptools. So, if
you're using data from a tool other than bamCoverage this is your starting point.</p>
<div class="highlight"><pre><span></span><code>computeMatrix scale-regions <span class="se">\</span>
    -p <span class="m">10</span> <span class="se">\</span>
    --startLabel start <span class="se">\</span>
    --endLabel end <span class="se">\</span>
    --upstream <span class="si">${</span><span class="nv">base_pairs</span><span class="si">}</span> <span class="se">\</span>
    --downstream <span class="si">${</span><span class="nv">base_pairs</span><span class="si">}</span> <span class="se">\</span>
    --regionBodyLength <span class="si">${</span><span class="nv">scale_length</span><span class="si">}</span> <span class="se">\</span>
    --regionsFileName <span class="si">${</span><span class="nv">regions_bed</span><span class="si">}</span> <span class="se">\</span>
    --scoreFileName <span class="si">${</span><span class="nv">input1_bigWig</span><span class="si">}</span> <span class="si">${</span><span class="nv">input2_bigWig</span><span class="si">}</span> <span class="se">\</span>
    --outFileName <span class="si">${</span><span class="nv">output_matrix</span><span class="p">.gz</span><span class="si">}</span>
</code></pre></div>


<p>The <code>--startLabel</code> and <code>--endLabel</code> values can be changed as desired, but don't
forget them! The <code>--upstream</code> and <code>--downstream</code> values can be as desired. The
<code>--regionBodyLength</code> is the value to which all features will be scaled. I suggest
using either the mean or median length of the features of interest. The regions
will be input as a .bed file, and the bigWig files that were generated in the
previous step will be used where indicated. Multiple files can be input,
space separated. You can specify that the matrix be gzipped by simply adding .gz
to the name of your output file. Now, the final step is to generate the plot and
also output the raw data:</p>
<div class="highlight"><pre><span></span><code>plotProfile <span class="se">\</span>
    --startLabel start <span class="se">\</span>
    --endLabel end <span class="se">\</span>
    --averageType <span class="si">${</span><span class="nv">mean</span><span class="p">|median</span><span class="si">}</span> <span class="se">\</span>
    --matrixFile <span class="si">${</span><span class="nv">input_matrix</span><span class="p">.gz</span><span class="si">}</span> <span class="se">\</span>
    --outFileName <span class="si">${</span><span class="nv">metaplot</span><span class="p">.svg</span><span class="si">}</span> <span class="se">\</span>
    --outFileNameData <span class="si">${</span><span class="nv">metaplot</span><span class="p">.tab</span><span class="si">}</span>
</code></pre></div>


<p>This will generate a plot, but also output the table of per-bin values that were
plotted. I made this with it:</p>
<p><center>
<img src="/figures/2019-06-27_making-better-metaplots-with-ggplot-part-1/deeptools_example_meta2.png", style="width:460px;height:275px;">
</center></p>
<p>I could play with Deeptools further, but the options for changing its aesthetics
are more limited than I'd like. In particular, smoothing the lines requires
smoothing the underlying data in the scoreMatrix step. Which I am not a huge fan
of. Now, let's load that table into R and make something prettier in
<a href="/articles/2019-06-28_making-better-metaplots-with-ggplot-part-2.html">Part 2</a>.</p>
  </div>
  <div class="tag-cloud">
    <p>
      <a href="https://groverj3.github.io/tag/bioinformatics.html">bioinformatics</a>
      <a href="https://groverj3.github.io/tag/data-visualization.html">data-visualization</a>
    </p>
  </div>





</article>

    <footer>
<p>
  &copy;  2019 - This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>
</p>
<p>    Powered by <a href="http://getpelican.com" target="_blank">Pelican</a> - <a href="https://github.com/alexandrevicenzi/flex" target="_blank">Flex</a> theme by <a href="http://alexandrevicenzi.com" target="_blank">Alexandre Vicenzi</a>
</p><p>
  <a rel="license"
     href="http://creativecommons.org/licenses/by-sa/4.0/"
     target="_blank">
    <img alt="Creative Commons License"
         title="Creative Commons License"
         style="border-width:0"
           src="https://i.creativecommons.org/l/by-sa/4.0/80x15.png"
         width="80"
         height="15"/>
  </a>
</p>    </footer>
  </main>




<script type="application/ld+json">
{
  "@context" : "http://schema.org",
  "@type" : "Blog",
  "name": " Jeff Grover. Bioinformatician. ",
  "url" : "https://groverj3.github.io",
  "image": "/images/jeff.jpg",
  "description": ""
}
</script>

</body>
</html>